source/MRMesh/MRPrecisePredicates2.cpp

```diff
@@ -43,6 +43,115 @@ bool ccw( const Vector2i & a, const Vector2i & b )
     return true;
 }
 
+bool orientParaboloid3d( const Vector2i & a0, const Vector2i & b0, const Vector2i & c0 )
+{
+    Vector3ll a( a0.x, a0.y, sqr( (long long) a0.x ) + sqr( (long long) a0.y ) );
+    Vector3ll b( b0.x, b0.y, sqr( (long long) b0.x ) + sqr( (long long) b0.y ) );
+    Vector3ll c( c0.x, c0.y, sqr( (long long) c0.x ) + sqr( (long long) c0.y ) );
+
+    using Vector2hp = Vector2<HighPrecisionInt>;
+    using Vector3hpp = Vector3<HighHighPrecisionInt>;
+
+    //e**0
+    if ( auto v = mixed( Vector3hpp{ a }, Vector3hpp{ b }, Vector3hpp{ c } ) )
+        return v > 0;
+
+    // e**1
+    const auto bxy_cxy = cross( Vector2hp{ b.x, b.y }, Vector2hp{ c.x, c.y } );
+    if ( auto v = -cross( Vector2hp{ b.x, b.z }, Vector2hp{ c.x, c.z } ) + 2 * a.y * bxy_cxy )
+        return v > 0;
+
+    // e**2
+    if ( auto v = bxy_cxy )
+        return v > 0;
+
+    // e**2
+    assert( bxy_cxy == 0 );
+    if ( auto v = cross( Vector2hp{ b.y, b.z }, Vector2hp{ c.y, c.z } ) ) // + 2 * a.x * bxy_cxy;
+        return v > 0;
+
+    // e**6 same as e**2
+
+    // e**9
+    const auto axy_cxy = cross( Vector2hp{ a.x, a.y }, Vector2hp{ c.x, c.y } );
+    if ( auto v = cross( Vector2hp{ a.x, a.z }, Vector2hp{ c.x, c.z } ) - 2 * b.y * axy_cxy )
+        return v > 0;
+
+    // e**10
+    if ( auto v = -c.x * ( a.y + b.y ) )
+        return v > 0;
+
+    // e**11
+    if ( auto v = -c.x )
+        return v > 0;
+
+    // e**12: -2*a.x*c.x - 2*b.y*c.y + c.z
+    assert( c.x == 0 );
+    if ( auto v = - 2 * b.y * c.y + c.z )
+        return v > 0;
+
+    // e**18
+    if ( auto v = -axy_cxy )
+        return v > 0;
+
+    // e**21
+    if ( auto v = -c.y )
+        return v > 0;
+    assert( c.x == 0 && c.y == 0 && c.z == 0 );
+
+    // e**81
+    if ( auto v = b.x * HighPrecisionInt( a.z ) - a.x * HighPrecisionInt( b.z ) )
+        return v > 0;
+
+    // e**82
+    if ( auto v = a.y * b.x )
+        return v > 0;
+
+    // e**83
+    if ( auto v = b.x )
+        return v > 0;
+
+    // e**84
+    if ( auto v = -b.z )
+        return v > 0; // can only be false, since b.z >= 0
+
+    // e**90
+    if ( auto v = -a.x * b.y )
+        return v > 0;
+
+    // e**93
+    if ( auto v = -b.y )
+        return v > 0;
```

Comment:
we should never fall here, because if `b.y` is not zero we will return in `e**84`